pkg/operator/staticpod/controller/monitoring/monitoring_resource_controller.go

```diff
@@ -0,0 +1,199 @@
+	dynamicClient        dynamic.Interface
+	operatorConfigClient common.OperatorClient
+	eventRecorder        events.Recorder
+}
+
+// NewMonitoringResourceController creates a new backing resource controller.
+func NewMonitoringResourceController(
+	targetNamespace string,
+	serviceMonitorName string,
+	operatorConfigClient common.OperatorClient,
+	kubeInformersForTargetNamespace informers.SharedInformerFactory,
+	kubeClient kubernetes.Interface,
+	dynamicClient dynamic.Interface,
+	eventRecorder events.Recorder,
+) *MonitoringResourceController {
+	c := &MonitoringResourceController{
+		targetNamespace:      targetNamespace,
+		operatorConfigClient: operatorConfigClient,
+		eventRecorder:        eventRecorder,
+		serviceMonitorName:   serviceMonitorName,
+
+		clusterRoleBindingListerSynced: kubeInformersForTargetNamespace.Core().V1().ServiceAccounts().Informer().HasSynced,
+		clusterRoleBindingLister:       kubeInformersForTargetNamespace.Rbac().V1().ClusterRoleBindings().Lister(),
+
+		queue:         workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "MonitoringResourceController"),
+		kubeClient:    kubeClient,
+		dynamicClient: dynamicClient,
+	}
+
+	operatorConfigClient.Informer().AddEventHandler(c.eventHandler())
+
+	// TODO: We need a dynamic informer here to observe changes to ServiceMonitor resource.
+
+	kubeInformersForTargetNamespace.Rbac().V1().ClusterRoleBindings().Informer().AddEventHandler(c.eventHandler())
+	return c
+}
+
+func (c MonitoringResourceController) mustTemplateAsset(name string) ([]byte, error) {
+	config := struct {
+		TargetNamespace string
+	}{
+		TargetNamespace: c.targetNamespace,
+	}
+	return assets.MustCreateAssetFromTemplate(name, bindata.MustAsset(filepath.Join(manifestDir, name)), config).Data, nil
+}
+
+func (c MonitoringResourceController) sync() error {
+	operatorSpec, _, _, err := c.operatorConfigClient.Get()
+	if err != nil {
+		return err
+	}
+
+	switch operatorSpec.ManagementState {
+	case operatorv1.Unmanaged:
+		return nil
+	case operatorv1.Removed:
+		// TODO: Should we try to actively remove the resources created by this controller here?
+		return nil
+	}
+
+	directResourceResults := resourceapply.ApplyDirectly(c.kubeClient, c.eventRecorder, c.mustTemplateAsset,
+		"manifests/prometheus-role.yaml",
+		"manifests/prometheus-role-binding.yaml",
+	)
+
+	errs := []error{}
+	for _, currResult := range directResourceResults {
+		if currResult.Error != nil {
+			errs = append(errs, fmt.Errorf("%q (%T): %v", currResult.File, currResult.Type, currResult.Error))
+		}
+	}
+
+	serviceMonitorBytes, err := c.mustTemplateAsset("manifests/service-monitor.yaml")
+	if err != nil {
+		errs = append(errs, fmt.Errorf("manifests/service-monitor.yaml: %v", err))
+	} else {
+		_, serviceMonitorErr := resourceapply.ApplyServiceMonitor(c.dynamicClient, c.eventRecorder, serviceMonitorBytes)
+		errs = append(errs, serviceMonitorErr)
+	}
+
+	err = common.NewMultiLineAggregate(errs)
```

Comment:
TODO could we refactor this actually just be a `common.ErrorToConditionMessage()` that handles the "normal" aggregate?

Reply:
https://github.com/openshift/library-go/issues/155